airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py

```diff
@@ -0,0 +1,126 @@
+from google.ads.googleads.v7.services.types.google_ads_service import SearchGoogleAdsResponse
+
+
+def chunk_date_range(start_date: str, end_date: str, conversion_window: Optional[int], field: str) -> Iterable[Mapping[str, any]]:
+    """
+    Passing optional parameter end_date for testing
+    Returns a list of the beginning and ending timetsamps of each month between the start date and now.
+    The return value is a list of dicts {'date': str} which can be used directly with the Slack API
+    """
+    intervals = []
+    end_date = date.fromisoformat(end_date) if end_date else date.today()
+    # As in to return some state when state in abnormal
+    if start_date > end_date.isoformat():
+        return [{field: start_date}]
+
+    # applying conversion windoe
+    start_date = date.fromisoformat(
+        start_date) - relativedelta(days=conversion_window)
+    yesterday = end_date - relativedelta(days=1)
+
+    # Each stream_slice contains the beginning and ending timestamp for a 24 hour period
+    while start_date < yesterday:
+        start = start_date
+        intervals.append({field: start.isoformat()})
+        start_date = min(yesterday, start_date + relativedelta(months=1))
+
+    return intervals
+
+
+class GoogleAdsStream(Stream, ABC):
+    def __init__(self, config):
+        self.config = config
+        self.google_ads_client = GoogleAds(**config)
+
+    def parse_response(self, response: SearchGoogleAdsResponse) -> Iterable[Mapping]:
+        for result in response:
+            record = GoogleAds.parse_single_result(
+                self.get_json_schema(), result)
+            yield record
+
+    def read_records(
+        self,
+        sync_mode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        start_date, end_date = Utils.get_date_params(
+            stream_slice, self.cursor_field)
+        query = GoogleAds.convert_schema_into_query(
+            self.get_json_schema(), self.name, start_date, end_date)
+
+        response = self.google_ads_client.send_request(
+            query)
+        yield from self.parse_response(response)
+
+class IncrementalGoogleAdsStream(GoogleAdsStream, ABC):
+    state_checkpoint_interval = None
+    CONVERSION_WINDOW_DAYS = 14
+
+    # Just to have to default value to check with latest_Record
+    TOO_OLD_DATE = "2000-01-01"
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        stream_state = stream_state or {}
+        start_date = stream_state.get(
+            self.cursor_field) or self.config.get("start_date")
+        return chunk_date_range(start_date, None, self.CONVERSION_WINDOW_DAYS, self.cursor_field)
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
```

Comment:
can we add a unit test for this method? I'm concerned that we are doing `max` on strings, not parsed date objects. Why don't we use something like `strptime` or even better use a library like `pendulum` to parse dates? Pendulum's iface is pretty user friendly